### ISSUE TITLE
Handle Access Denied Errors Differently

### ISSUE THREAD

**Issue by cmd-ctrl-freq** [turbot/steampipe-plugin-aws#878]
**Is your feature request related to a problem? Please describe.**
I'm writing controls for a custom AWS security mod. I want my controls to check every resource, but when my connection is denied access to a subset of resources for a given table, the query corresponding to my control fails altogether.

**Describe the solution you'd like**
For my use case, dropping the row for the resource to which access is denied is suitable. (see below for problems associated with this solution)

**Describe alternatives you've considered**
Alternatively, returning a null row for the resource to which access is denied would also work. I've proposed a "status" column which would contain API responses to give insight into issues. (see below for problems associated with this solution)

**Comment by cmd-ctrl-freq** [comment]
Relevant Slack Conversation for context:
[Mitch Fentz]
So, I'm writing a control that checks a specific AWS resource (kms keys in this case). On one of those resources, my access is denied. Because of this, the entire query fails. Is anyone aware of a way to work around this?
I suppose that my preferred behavior would be for the AWS plugin to just skip over that resource, and continue building out the table.

[Mitch Fentz]
Alternatively, AWS resources could have a status column. Resources which are accessible could be populated as normal, with the response code from the API call in the status column, and unreachable resources could be populated with null values, with the response code in the status column.

[Jon Udell]
Interesting ideas, thanks [@Mitch Fentz] [@Cody]/ [@John Smyth] what do you think?

[Ryan Gerstenkorn]
Yeah, this makes it difficult to work around when you have partial access, for the most part I would say it makes sense to just assign the right permissions but this can also happen when you have say a locked down bucket which is restricted by resource policies. Resource policies are typically more difficult to get updated then IAM permissions, especially when there is more then a few. (edited) 

[John Smyth]
Others have run into similar issues with unauthorized calls. I don't really like the idea of returning 'null' because it already means something - you don't know that it is null because the call was unauthorized.  We could add a 'status' column as was suggested, but it would actually have to be a jsonb object that said which columns were unauthorized, as a given table can have multiple hydrate functions, and each can populate multiple columns.  Also, it wouldn't solve the case where the list call itself is unauthorized for a give region or account.  Finally, it would require the user to actually correlate that null with that status/reason.
I think the fundamental challenges are:
`null`  already means something.  While we could track the 'reason' elsewhere, it's not obvious that the data is missing because it was unauthorized, not because the api returned null.  Your control may 'pass' even though the resource does not comply, because you were unauthorized.
 There's no way to notify you when it happens without throwing an error.  WHEN your control 'passes' even though the resource does not comply, you don't know that its because you were unauthorized.
 We don't always know your 'intent' - maybe you want to ignore this one because you know you are not authorized, but we may also ignore other unauthorized resources, columns, etc that you were unaware of.
I think our current approach is the most correct from the data accuracy perspective.  I think that we need to explore our options though, and I appreciate this (and any other) suggestions.

**Comment by rajlearner17** [comment]
@cmd-ctrl-freq Appreciate using Steampipe. Thanks for raising this issue!!
We are looking into it along with @johnsmyth 

[slack reference ](https://steampipe.slack.com/archives/C01UECB59A7/p1643840445543609)

**Comment by johnsmyth** [comment]
I think we should move ahead with @e-gineer's proposal from the slack thread:
> My suggestion would be an option in the AWS plugin itself. e.g. `ignore_access_denied_errors` = true . If set, then the plugin would choose to skip access denied problems or return partial data for them. The default of the option would be false. Then ... for advanced users like [@Mitch Fentz](https://steampipe.slack.com/team/U02SW91F9DY) or [@Ryan Gerstenkorn](https://steampipe.slack.com/team/U01UMBY1YKH) hitting this problem they could choose to make it go away ... but for new users they would be warned of the problems / access denied errors.

I think we will need @kaidaguerre to add a `ShouldIgnoreError` capability for hydrate functions to the SDK, and then we (@cbruno10 ?  @LalitTurbot) can implement `ignore_access_denied_errors` for the AWS plugin.  I don't think that we should consider this settled precedent though; we should only implement for the AWS plugin now, and discuss other situations as they arise - this is a pretty blunt instrument.

**Comment by cbruno10** [comment]
@johnsmyth I believe there's already a way to define a `ShouldIgnoreError` for a hydrate function, e.g., https://github.com/turbot/steampipe-plugin-chaos/blob/main/chaos/table_chaos_hydrate_errors.go#L30-L38. @LalitTurbot is currently looking at how to implement this for the `aws_kms_key` table, and others, in the AWS plugin in a scalable way.

**Comment by cbruno10** [comment]
Our current approach is to add a new connection config argument (currently named `ignored_errror_codes`, name not finalized) that would accept a list of error codes as string that the AWS plugin would ignore, e.g.,

```
connection "aws" {
  plugin    = "aws"
  profile   = "my-profile"
  regions = ["us-east-1"]

  # Ignore common unauthorized error codes to prevent failures
  ignored_error_codes = ["AccessDenied", "AccessDeniedException", "NotAuthorized", "UnauthorizedOperation", "UnrecognizedClientException"]
}
```

The default value would be the most common unauthorized error codes across all services and resource types, and the list could be expanded with any additional error codes (of any type).

These errors would be ignored, so if any queries would hit these errors, instead of causing execution to fail, an empty row would be returned (for an entire resource) or the column would be set to `null`. 

Any errors that tables ignore today, like common not found errors, would still be ignored, regardless of what is specified in the `ignored_error_codes` config argument.

@cmd-ctrl-freq - Any thoughts or feedback on this approach? I'm interested to hear if this would help your original use case, or any others you have in mind. Thanks!

**Comment by cmd-ctrl-freq** [comment]
@cbruno10 I think this is a pretty good solution. I like the fact that this way I can specify which errors I want to ignore. 

Is this argument searching for substrings in the error message? Like
`if  "AccessDenied" in error.message then ignore`

Initially I was thinking that it would be nice to have an "all" or "*" argument to ignore everything, but now that I've thought about it I don't think I would end up using it. Might still be working adding as an option though.

**Comment by cbruno10** [comment]
@cmd-ctrl-freq Yes, I think supporting wildcards like `*` and `?` is feasible, we'll do some testing with them and see if anything unexpected pops up.

Are there any common AWS error code patterns you anticipate using to ignore for all resource types?

**Comment by cbruno10** [comment]
Hi @cmd-ctrl-freq , this feature has been released as part of v0.60.0. For more information on the `ignore_error_codes` config arg, please see https://hub.steampipe.io/plugins/turbot/aws#configuration.
